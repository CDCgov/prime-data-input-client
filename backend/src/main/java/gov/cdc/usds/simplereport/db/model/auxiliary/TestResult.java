package gov.cdc.usds.simplereport.db.model.auxiliary;

public enum TestResult {
	POSITIVE, NEGATIVE, UNDETERMINED
}
