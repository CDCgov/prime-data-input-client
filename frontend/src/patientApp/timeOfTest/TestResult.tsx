import { useSelector } from "react-redux";

import { formatFullName } from "../../app/utils/user";
import { RootState } from "../../app/store";
import { Patient } from "../../app/patients/ManagePatients";
import { TestResult as TestResultType } from "../../app/testQueue/QueueItem";
import { COVID_RESULTS, TEST_RESULT_DESCRIPTIONS } from "../../app/constants";

const TestResult = () => {
  const patient = useSelector<RootState, Patient>((state) => state.patient);
  const fullName = formatFullName(patient as any);
  const dateTested = new Date(patient.lastTest.dateTested).toLocaleDateString();
  const deviceType = patient.lastTest.deviceTypeModel;

  return (
    <main className="patient-app padding-top-105 padding-bottom-4 bg-base-lightest">
      <div className="grid-container maxw-tablet">
        <div className="card usa-prose">
          <h1 className="font-heading-lg">SARS-CoV-2 result</h1>
          <h2 className="font-heading-sm">Patient</h2>
          <p className="margin-top-05">{fullName}</p>
          <div className="grid-row">
            <div className="grid-col usa-prose">
              <h2 className="font-heading-sm">Test result</h2>
              <p className="margin-top-05">
                {TEST_RESULT_DESCRIPTIONS[patient.lastTest.result]}
              </p>
            </div>
            <div className="grid-col usa-prose">
              <h2 className="font-heading-sm">Test date</h2>
              <p className="margin-top-05">{dateTested}</p>
            </div>
          </div>
          <h2 className="font-heading-sm">Test device</h2>
          <p className="margin-top-05">{deviceType}</p>
          <h2 className="font-heading-sm">What does my result mean?</h2>
          <TestResultNotes result={patient.lastTest.result} />
          <p>
            For more information, please visit the{" "}
            <a href="https://www.cdc.gov/coronavirus/2019-ncov/if-you-are-sick/end-home-isolation.html">
              Centers for Disease Control and Prevention (CDC) website
            </a>{" "}
            or contact your local health department.
          </p>
        </div>
      </div>
    </main>
  );
};

interface TestResultNotesProps {
  result: TestResultType;
}

const TestResultNotes: React.FC<TestResultNotesProps> = (props) => {
  switch (props.result) {
    case COVID_RESULTS.POSITIVE:
      return (
        <>
          <p>
            Getting a positive COVID-19 test result can be difficult news, so
            it’s important to{" "}
            <a href="https://www.cdc.gov/coronavirus/2019-ncov/daily-life-coping/managing-stress-anxiety.html">
              take steps to cope with stress
            </a>{" "}
            during this time. Reach out to your support system and make a phone
            or video appointment with a mental health professional if needed.
          </p>
          <p>
            Most people who get COVID-19 will be able to recover at home. Make
            sure to follow CDC guidelines for people who are recovering at home
            and their caregivers, including:
          </p>
          <ul>
            <li>Stay home when you are sick, except to get medical care.</li>
            <li>
              Self isolate for 10 full days after symptoms first appeared (or
              starting the day after you had your test, if you have no
              symptoms).
            </li>
            <li>
              If you are self isolating at home where others live, use a
              separate room and bathroom for sick household members (if
              possible). Clean any shared rooms as needed, to avoid transmitting
              the virus.
            </li>
            <li>
              Wash your hands often with soap and water for at least 20 seconds,
              especially after blowing your nose, coughing, or sneezing; going
              to the bathroom; and before eating or preparing food.
            </li>
            <li>
              If soap and water are not available, use an alcohol-based hand
              sanitizer with at least 60% alcohol.
            </li>
            <li>
              Have a supply of clean, disposable face masks. Everyone, no matter
              their COVID-19 diagnosis, should wear face masks while in the
              home.
            </li>
          </ul>
          <p>
            <a href="https://www.cdc.gov/coronavirus/2019-ncov/symptoms-testing/symptoms.html">
              Watch for symptoms and learn when to seek emergency medical
              attention
            </a>
            .
          </p>
          <p>
            If someone is showing any of these signs, seek emergency medical
            care immediately:
          </p>
          <ul>
            <li>Trouble breathing</li>
            <li>Persistent chest pain/pressure</li>
            <li>Confusion</li>
            <li>Inability to wake or stay awake</li>
            <li>Bluish lips or face</li>
          </ul>
          <p>
            Call 911 or call ahead to your local emergency room: Notify the
            operator that you are seeking care for someone who has or may have
            COVID-19.
          </p>
        </>
      );
    case COVID_RESULTS.NEGATIVE:
      return (
        <>
          <p>
            COVID-19 antigen tests can sometimes provide inaccurate or false
            results and follow up testing may be needed. Continue social
            distancing and wearing a mask. Contact your healthcare provider to
            decide if additional testing is needed, especially if you experience
            any of these symptoms:
          </p>
          <ul>
            <li>Fever or chills</li>
            <li>Cough</li>
            <li>Shortness of breath or difficulty breathing</li>
            <li>Fatigue</li>
            <li>Muscle or body aches</li>
            <li>Headache</li>
            <li>Loss of taste or smell</li>
            <li>Sore throat</li>
            <li>Congestion or runny nose</li>
            <li>Nausea or vomiting</li>
            <li>Diarrhea</li>
          </ul>
        </>
      );
    default:
      return (
        <>
          <p>
            An inconclusive result is neither positive nor negative. This can
            happen because of problems with the sample collection, a very
            early-stage COVID-19 infection, or for patients with COVID-19 that
            are close to recovery. With an inconclusive result, collecting and
            testing another sample is recommended.
          </p>
          <p>
            Please make an appointment for another test as soon as possible. If
            you’ve gotten tested due to COVID-19 symptoms, it is recommended
            that you self-isolate until you get your new test results.
          </p>
        </>
      );
  }
};

export default TestResult;
