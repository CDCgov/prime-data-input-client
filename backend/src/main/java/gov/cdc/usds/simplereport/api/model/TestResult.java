package gov.cdc.usds.simplereport.api.model;

import java.util.Date;
import java.util.Map;
import java.time.LocalDate;

import org.json.JSONObject;

import gov.cdc.usds.simplereport.db.model.Person;
import gov.cdc.usds.simplereport.db.model.Organization;
import gov.cdc.usds.simplereport.db.model.TestEvent;
import gov.cdc.usds.simplereport.db.model.DeviceType;
import gov.cdc.usds.simplereport.db.model.auxiliary.AskOnEntrySurvey;


public class TestResult {

	private TestEvent event;
	private AskOnEntrySurvey survey;

	public TestResult(TestEvent event) {
		super();
		this.event = event;
		this.survey = event.getTestOrder().getAskOnEntrySurvey().getSurvey();

	}

	public String getInternalId() {
		return event.getInternalId().toString();
	}

	public Organization getOrganization() {
		return event.getOrganization();
	}

	public Date getDateAdded() {
		return event.getTestOrder().getDateAdded();
	}

	public String getPregnancy() {
		return survey.getPregnancy();
	}

	public Boolean getNoSymptoms() {
		return survey.getNoSymptoms();
	}

	public String getSymptoms() {
		Map<String, Boolean> s = survey.getSymptoms();
		JSONObject obj = new JSONObject();
		for (Map.Entry<String,Boolean> entry : s.entrySet()) {
			obj.put(entry.getKey(), entry.getValue().toString());
		}
		return obj.toString();
	}

	public LocalDate getSymptomOnset() {
		return survey.getSymptomOnsetDate();
	}


	public Boolean getFirstTest() {
		return survey.getFirstTest();
	}


	public LocalDate getPriorTestDate() {
		return survey.getPriorTestDate();
	}


	public String getPriorTestType() {
		return survey.getPriorTestType();
	}


	public String getPriorTestResult() {
		return survey.getPriorTestResult() == null ? "" : survey.getPriorTestResult().toString();
	}


	public DeviceType getDeviceType() {
		return event.getDeviceType();
	}


	public Person getPatient() {
		return event.getPatient();
	}


	public String getResult() {
		return event.getResult().toString();
	}


	public Date getDateTested() {
		return event.getCreatedAt();
	}


}
