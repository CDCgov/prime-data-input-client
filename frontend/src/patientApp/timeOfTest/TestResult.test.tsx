import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import { Router } from "react-router-dom";
import { createMemoryHistory } from "history";
import { render } from "@testing-library/react";

import TestResult from "./TestResult";

const mockStore = configureStore([]);

describe("TestResult", () => {
  it("should show a positive result", () => {
    const store = mockStore({
      plid: "foo",
      patient: {
        firstName: "Bob",
        lastName: "Barker",
        lastTest: {
          deviceType: "Testing device",
          dateTested: "08/27/2020",
          result: "POSITIVE",
        },
      },
    });
    const { container, getByText } = render(
      <Router history={createMemoryHistory()}>
        <Provider store={store}>
          <TestResult />
        </Provider>
      </Router>
    );

    expect(getByText("SARS-CoV-2 result")).toBeInTheDocument();
    expect(getByText("Bob Barker")).toBeInTheDocument();
    expect(getByText("Test result")).toBeInTheDocument();
    expect(getByText("Positive")).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
  it("should show a negative result", () => {
    const store = mockStore({
      plid: "foo",
      patient: {
        firstName: "Bob",
        lastName: "Barker",
        lastTest: {
          deviceType: "Testing device",
          dateTested: "08/27/2020",
          result: "NEGATIVE",
        },
      },
    });
    const { container, getByText } = render(
      <Router history={createMemoryHistory()}>
        <Provider store={store}>
          <TestResult />
        </Provider>
      </Router>
    );

    expect(getByText("SARS-CoV-2 result")).toBeInTheDocument();
    expect(getByText("Bob Barker")).toBeInTheDocument();
    expect(getByText("Test result")).toBeInTheDocument();
    expect(getByText("Negative")).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
  it("should show an inconclusive result", () => {
    const store = mockStore({
      plid: "foo",
      patient: {
        firstName: "Bob",
        lastName: "Barker",
        lastTest: {
          deviceType: "Testing device",
          dateTested: "08/27/2020",
          result: "UNDETERMINED",
        },
      },
    });
    const { container, getByText } = render(
      <Router history={createMemoryHistory()}>
        <Provider store={store}>
          <TestResult />
        </Provider>
      </Router>
    );

    expect(getByText("SARS-CoV-2 result")).toBeInTheDocument();
    expect(getByText("Bob Barker")).toBeInTheDocument();
    expect(getByText("Test result")).toBeInTheDocument();
    expect(getByText("Inconclusive")).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
