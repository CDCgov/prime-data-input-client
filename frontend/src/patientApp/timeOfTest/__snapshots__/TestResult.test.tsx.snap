// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TestResult should show a negative result 1`] = `
<div>
  <main
    class="patient-app padding-top-105 padding-bottom-4 bg-base-lightest"
  >
    <div
      class="grid-container maxw-tablet"
    >
      <div
        class="card usa-prose"
      >
        <h1
          class="font-heading-lg"
        >
          SARS-CoV-2 result
        </h1>
        <h2
          class="font-heading-sm"
        >
          Patient
        </h2>
        <p
          class="margin-top-05"
        >
          Bob Barker
        </p>
        <div
          class="grid-row"
        >
          <div
            class="grid-col usa-prose"
          >
            <h2
              class="font-heading-sm"
            >
              Test result
            </h2>
            <p
              class="margin-top-05"
            >
              Negative
            </p>
          </div>
          <div
            class="grid-col usa-prose"
          >
            <h2
              class="font-heading-sm"
            >
              Test date
            </h2>
            <p
              class="margin-top-05"
            >
              8/27/2020
            </p>
          </div>
        </div>
        <h2
          class="font-heading-sm"
        >
          Test device
        </h2>
        <p
          class="margin-top-05"
        >
          Testing device
        </p>
        <h2
          class="font-heading-sm"
        >
          Notes
        </h2>
        <p>
          Antigen tests can return inaccurate or false results and follow up testing may be needed. Continue social distancing and wearing a mask. Contact your healthcare provider to determine if additional testing is needed especially if you experience any of these COVID-19 symptoms:
        </p>
        <ul>
          <li>
            Fever or chills
          </li>
          <li>
            Cough
          </li>
          <li>
            Shortness of breath or difficulty breathing
          </li>
          <li>
            Fatigue
          </li>
          <li>
            Muscle or body aches
          </li>
          <li>
            Headache
          </li>
          <li>
            New loss of taste or smell
          </li>
          <li>
            Sore throat
          </li>
          <li>
            Congestion or runny nose
          </li>
          <li>
            Nausea or vomiting
          </li>
          <li>
            Diarrhea
          </li>
        </ul>
        <p>
          For further guidance, please consult the
           
          <a
            href="https://www.cdc.gov/coronavirus/2019-ncov/if-you-are-sick/end-home-isolation.html"
          >
            Centers for Disease Control and Prevention (CDC)
          </a>
           
          website or contact your local health department.
        </p>
      </div>
    </div>
  </main>
</div>
`;

exports[`TestResult should show a positive result 1`] = `
<div>
  <main
    class="patient-app padding-top-105 padding-bottom-4 bg-base-lightest"
  >
    <div
      class="grid-container maxw-tablet"
    >
      <div
        class="card usa-prose"
      >
        <h1
          class="font-heading-lg"
        >
          SARS-CoV-2 result
        </h1>
        <h2
          class="font-heading-sm"
        >
          Patient
        </h2>
        <p
          class="margin-top-05"
        >
          Bob Barker
        </p>
        <div
          class="grid-row"
        >
          <div
            class="grid-col usa-prose"
          >
            <h2
              class="font-heading-sm"
            >
              Test result
            </h2>
            <p
              class="margin-top-05"
            >
              Positive
            </p>
          </div>
          <div
            class="grid-col usa-prose"
          >
            <h2
              class="font-heading-sm"
            >
              Test date
            </h2>
            <p
              class="margin-top-05"
            >
              8/27/2020
            </p>
          </div>
        </div>
        <h2
          class="font-heading-sm"
        >
          Test device
        </h2>
        <p
          class="margin-top-05"
        >
          Testing device
        </p>
        <h2
          class="font-heading-sm"
        >
          Notes
        </h2>
        <p>
          Please self-isolate at home. You can be around others after:
        </p>
        <ul>
          <li>
            10 days since symptoms first appeared 
            <b>
              and
            </b>
          </li>
          <li>
            24 hours with no fever without the use of fever-reducing medications 
            <b>
              and
            </b>
          </li>
          <li>
            Other symptoms of COVID-19 are improving*
          </li>
        </ul>
        <p>
          *Loss of taste and smell may persist for weeks or months after recovery and need not delay the end of isolationâ€‹
        </p>
        <p>
          Most people do not require testing to decide when they can be around others; however, if your healthcare provider recommends testing, they will let you know when you can resume being around others based on your test results.
        </p>
        <p>
          Note that these recommendations 
          <b>
            do not
          </b>
           apply to persons with severe COVID-19 or with severely weakened immune systems (immunocompromised).
        </p>
        <p>
          For further guidance, please consult the
           
          <a
            href="https://www.cdc.gov/coronavirus/2019-ncov/if-you-are-sick/end-home-isolation.html"
          >
            Centers for Disease Control and Prevention (CDC)
          </a>
           
          website or contact your local health department.
        </p>
      </div>
    </div>
  </main>
</div>
`;

exports[`TestResult should show an inconclusive result 1`] = `
<div>
  <main
    class="patient-app padding-top-105 padding-bottom-4 bg-base-lightest"
  >
    <div
      class="grid-container maxw-tablet"
    >
      <div
        class="card usa-prose"
      >
        <h1
          class="font-heading-lg"
        >
          SARS-CoV-2 result
        </h1>
        <h2
          class="font-heading-sm"
        >
          Patient
        </h2>
        <p
          class="margin-top-05"
        >
          Bob Barker
        </p>
        <div
          class="grid-row"
        >
          <div
            class="grid-col usa-prose"
          >
            <h2
              class="font-heading-sm"
            >
              Test result
            </h2>
            <p
              class="margin-top-05"
            >
              Inconclusive
            </p>
          </div>
          <div
            class="grid-col usa-prose"
          >
            <h2
              class="font-heading-sm"
            >
              Test date
            </h2>
            <p
              class="margin-top-05"
            >
              8/27/2020
            </p>
          </div>
        </div>
        <h2
          class="font-heading-sm"
        >
          Test device
        </h2>
        <p
          class="margin-top-05"
        >
          Testing device
        </p>
        <h2
          class="font-heading-sm"
        >
          Notes
        </h2>
        <p>
          An inconclusive result is neither positive nor negative. This result can occur from inadequate sample collection, very early-stage infection, or for patients close to recovery. With an inconclusive result, collecting and testing another sample is recommended. Please make an appointment for another test as soon as possible.
        </p>
        <p>
          For further guidance, please consult the
           
          <a
            href="https://www.cdc.gov/coronavirus/2019-ncov/if-you-are-sick/end-home-isolation.html"
          >
            Centers for Disease Control and Prevention (CDC)
          </a>
           
          website or contact your local health department.
        </p>
      </div>
    </div>
  </main>
</div>
`;
